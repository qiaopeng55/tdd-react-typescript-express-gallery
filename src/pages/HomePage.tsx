import { MenuItem, Select, Typography } from "@material-ui/core";

import { withStyles } from "@material-ui/core/styles";

import { Col, Row } from "react-flexbox-grid";

import Pagination from "material-ui-flat-pagination";

import * as React from "react";
import { connect } from "react-redux";

import { Product } from "../model/model";
import { IRootState } from "../reducers";

import { IPaginate } from "../actions/constant";
import { paginateProducts } from "../actions/product";

import {
  PaginationSetting,
  paginationStyles,
  WhiteSpace,
  Divider
} from "../components/StyledComponents";

import ProductContainer from "../components/ProductContainer";

export interface IProps {
  classes: any;
  products: Product[];
  product_total: number;
  paginateProducts: ({ start, limit }: IPaginate) => void;
}

export class HomePage extends React.Component<IProps> {
  public state = {
    limit: 8,
    offset: 0
  };

  public handleLimitChange = (event: any) => {
    this.setState({
      offset: 0,
      limit: event.target.value
    });

    this.props.paginateProducts({
      start: 0,
      limit: event.target.value
    });
  };

  public handlePagination(offset: number) {
    this.setState({ offset });

    this.props.paginateProducts({
      start: offset,
      limit: this.state.limit
    });
  }

  public componentDidMount() {
    this.props.paginateProducts({
      start: this.state.offset,
      limit: this.state.limit
    });
  }

  public render() {
    const { classes, product_total, products } = this.props;

    const { paperRoot, ...paginationClasses } = classes;

    const paperClasses = {
      root: paperRoot
    };

    const Products = products.map((product: Product) => (
      <ProductContainer key={product.id} product={product} />
    ));

    return (
      <div>
        <Typography id="title" variant="h5">
          All Products
        </Typography>

        <PaginationSetting>
          <div>
            <Typography id="product_total" variant="caption">
              {`${product_total} Products`}
            </Typography>
          </div>
          <div>
            <Select
              disableUnderline={true}
              value={this.state.limit}
              name="limit"
              onChange={this.handleLimitChange}
            >
              <MenuItem value={8}>8 per page</MenuItem>
              <MenuItem value={16}>16 per page</MenuItem>
              <MenuItem value={20}>20 per page</MenuItem>
            </Select>
          </div>
        </PaginationSetting>

        <Divider weight={1} />

        <Row>{Products}</Row>

        <WhiteSpace height={30} />
        <Row end="xs">
          <Pagination
            limit={this.state.limit}
            offset={this.state.offset}
            total={product_total}
            classes={paginationClasses}
            currentPageColor="inherit"
            previousPageLabel={"< Previous page"}
            nextPageLabel={"Next page >"}
            onClick={(e, offset) => this.handlePagination(offset)}
            otherPageColor="inherit"
            reduced={true}
          />
        </Row>
      </div>
    );
  }
}

function mapStateToProps(state: IRootState) {
  return {
    products: state.products,
    product_total: state.pagination.product_total
  };
}

export default withStyles(paginationStyles)(
  connect(
    mapStateToProps,
    {
      paginateProducts
    }
  )(HomePage)
);
