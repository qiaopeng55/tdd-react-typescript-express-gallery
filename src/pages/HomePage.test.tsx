import { shallow } from "enzyme";
import * as React from "react";
import { HomePage } from "./HomePage";

const product_total = 1000;
const paginateProducts = jest.fn();

const props = {
  classes: { paperRoot: undefined },
  products: [],
  product_total,
  paginateProducts
};
describe("HomePage", () => {
  const wrap = shallow(<HomePage {...props} />);

  it("renders properly", () => {
    expect(wrap).toMatchSnapshot();
  });

  it("renders the title", () => {
    expect(
      wrap
        .find("#title")
        .children()
        .text()
    ).toEqual("All Products");
  });

  it("renders the correct products total", () => {
    expect(
      wrap
        .find("#product_total")
        .children()
        .text()
    ).toEqual(`${product_total} Products`);
  });

  it("calls the paginateProducts callback", () => {
    expect(paginateProducts).toHaveBeenCalled();
  });

  describe("limit selector", () => {
    it("renders 3 options for limit selector", () => {
      expect(wrap.find("WithStyles(Select)").children().length).toEqual(3);
    });

    it("selects 8 as the default limit", () => {
      expect(wrap.find("WithStyles(Select)").props()["value"]).toEqual(8);
    });

    describe("when changes limit value", () => {
      const limit = 16;

      beforeEach(() => {
        wrap
          .find("WithStyles(Select)")
          .simulate("change", { target: { value: limit } });
      });

      it("changes this.state.limit value", () => {
        expect(wrap.state("limit")).toEqual(limit);
      });

      it("calls the paginateProducts callback with expected parameters", () => {
        const expectedParams = { limit, start: 0 };

        expect(paginateProducts).toHaveBeenCalledWith(expectedParams);
      });
    });
  });

  it("renders pagination buttons", () => {
    expect(wrap.find("WithStyles(Pagination)").exists()).toBeTruthy();
  });

  describe("when clicks pagination button", () => {
    beforeEach(() => {
      wrap.find("WithStyles(Pagination)").simulate("click");
    });

    it("calls the paginateProducts callback", () => {
      expect(paginateProducts).toHaveBeenCalledWith({
        limit: 16,
        start: undefined
      });
    });
  });
});
