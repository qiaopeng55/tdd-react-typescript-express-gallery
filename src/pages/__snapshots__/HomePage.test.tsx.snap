// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <HomePage
    classes={
      Object {
        "paperRoot": undefined,
      }
    }
    paginateProducts={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "limit": 8,
              "start": 0,
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    product_total={1000}
    products={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <WithStyles(Typography)
          id="title"
          variant="h5"
        >
          All Products
        </WithStyles(Typography)>,
        <ForwardRef>
          <div>
            <WithStyles(Typography)
              id="product_total"
              variant="caption"
            >
              1000 Products
            </WithStyles(Typography)>
          </div>
          <div>
            <WithStyles(Select)
              disableUnderline={true}
              name="limit"
              onChange={[Function]}
              value={8}
            >
              <WithStyles(MenuItem)
                value={8}
              >
                8 per page
              </WithStyles(MenuItem)>
              <WithStyles(MenuItem)
                value={16}
              >
                16 per page
              </WithStyles(MenuItem)>
              <WithStyles(MenuItem)
                value={20}
              >
                20 per page
              </WithStyles(MenuItem)>
            </WithStyles(Select)>
          </div>
        </ForwardRef>,
        <ForwardRef
          weight={1}
        />,
        <Row />,
        <ForwardRef
          height={30}
        />,
        <Row
          end="xs"
        >
          <WithStyles(Pagination)
            classes={Object {}}
            currentPageColor="inherit"
            limit={8}
            nextPageLabel="Next page >"
            offset={0}
            onClick={[Function]}
            otherPageColor="inherit"
            previousPageLabel="< Previous page"
            reduced={true}
            total={1000}
          />
        </Row>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "All Products",
          "id": "title",
          "variant": "h5",
        },
        "ref": null,
        "rendered": "All Products",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <div>
              <WithStyles(Typography)
                id="product_total"
                variant="caption"
              >
                1000 Products
              </WithStyles(Typography)>
            </div>,
            <div>
              <WithStyles(Select)
                disableUnderline={true}
                name="limit"
                onChange={[Function]}
                value={8}
              >
                <WithStyles(MenuItem)
                  value={8}
                >
                  8 per page
                </WithStyles(MenuItem)>
                <WithStyles(MenuItem)
                  value={16}
                >
                  16 per page
                </WithStyles(MenuItem)>
                <WithStyles(MenuItem)
                  value={20}
                >
                  20 per page
                </WithStyles(MenuItem)>
              </WithStyles(Select)>
            </div>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <WithStyles(Typography)
                id="product_total"
                variant="caption"
              >
                1000 Products
              </WithStyles(Typography)>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "1000 Products",
                "id": "product_total",
                "variant": "caption",
              },
              "ref": null,
              "rendered": "1000 Products",
              "type": [Function],
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <WithStyles(Select)
                disableUnderline={true}
                name="limit"
                onChange={[Function]}
                value={8}
              >
                <WithStyles(MenuItem)
                  value={8}
                >
                  8 per page
                </WithStyles(MenuItem)>
                <WithStyles(MenuItem)
                  value={16}
                >
                  16 per page
                </WithStyles(MenuItem)>
                <WithStyles(MenuItem)
                  value={20}
                >
                  20 per page
                </WithStyles(MenuItem)>
              </WithStyles(Select)>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  <WithStyles(MenuItem)
                    value={8}
                  >
                    8 per page
                  </WithStyles(MenuItem)>,
                  <WithStyles(MenuItem)
                    value={16}
                  >
                    16 per page
                  </WithStyles(MenuItem)>,
                  <WithStyles(MenuItem)
                    value={20}
                  >
                    20 per page
                  </WithStyles(MenuItem)>,
                ],
                "disableUnderline": true,
                "name": "limit",
                "onChange": [Function],
                "value": 8,
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": "8 per page",
                    "value": 8,
                  },
                  "ref": null,
                  "rendered": "8 per page",
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": "16 per page",
                    "value": 16,
                  },
                  "ref": null,
                  "rendered": "16 per page",
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": "20 per page",
                    "value": 20,
                  },
                  "ref": null,
                  "rendered": "20 per page",
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
            "type": "div",
          },
        ],
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": undefined,
          "componentStyle": ComponentStyle {
            "componentId": "sc-bwzfXH",
            "isStatic": true,
            "rules": Array [
              "
  display: flex;
  justify-content: space-between;
  align-items: center;
",
            ],
          },
          "displayName": "styled.div",
          "render": [Function],
          "styledComponentId": "sc-bwzfXH",
          "target": "div",
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "weight": 1,
        },
        "ref": null,
        "rendered": null,
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": undefined,
          "componentStyle": ComponentStyle {
            "componentId": "sc-bxivhb",
            "isStatic": false,
            "rules": Array [
              "
  border: ",
              [Function],
              "px solid #eeeeee;
",
            ],
          },
          "displayName": "styled.div",
          "render": [Function],
          "styledComponentId": "sc-bxivhb",
          "target": "div",
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [],
        },
        "ref": null,
        "rendered": Array [],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "height": 30,
        },
        "ref": null,
        "rendered": null,
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": undefined,
          "componentStyle": ComponentStyle {
            "componentId": "sc-htpNat",
            "isStatic": false,
            "rules": Array [
              "
  height: ",
              [Function],
              "px;
",
            ],
          },
          "displayName": "styled.div",
          "render": [Function],
          "styledComponentId": "sc-htpNat",
          "target": "div",
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <WithStyles(Pagination)
            classes={Object {}}
            currentPageColor="inherit"
            limit={8}
            nextPageLabel="Next page >"
            offset={0}
            onClick={[Function]}
            otherPageColor="inherit"
            previousPageLabel="< Previous page"
            reduced={true}
            total={1000}
          />,
          "end": "xs",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "classes": Object {},
            "currentPageColor": "inherit",
            "limit": 8,
            "nextPageLabel": "Next page >",
            "offset": 0,
            "onClick": [Function],
            "otherPageColor": "inherit",
            "previousPageLabel": "< Previous page",
            "reduced": true,
            "total": 1000,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <WithStyles(Typography)
            id="title"
            variant="h5"
          >
            All Products
          </WithStyles(Typography)>,
          <ForwardRef>
            <div>
              <WithStyles(Typography)
                id="product_total"
                variant="caption"
              >
                1000 Products
              </WithStyles(Typography)>
            </div>
            <div>
              <WithStyles(Select)
                disableUnderline={true}
                name="limit"
                onChange={[Function]}
                value={8}
              >
                <WithStyles(MenuItem)
                  value={8}
                >
                  8 per page
                </WithStyles(MenuItem)>
                <WithStyles(MenuItem)
                  value={16}
                >
                  16 per page
                </WithStyles(MenuItem)>
                <WithStyles(MenuItem)
                  value={20}
                >
                  20 per page
                </WithStyles(MenuItem)>
              </WithStyles(Select)>
            </div>
          </ForwardRef>,
          <ForwardRef
            weight={1}
          />,
          <Row />,
          <ForwardRef
            height={30}
          />,
          <Row
            end="xs"
          >
            <WithStyles(Pagination)
              classes={Object {}}
              currentPageColor="inherit"
              limit={8}
              nextPageLabel="Next page >"
              offset={0}
              onClick={[Function]}
              otherPageColor="inherit"
              previousPageLabel="< Previous page"
              reduced={true}
              total={1000}
            />
          </Row>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "All Products",
            "id": "title",
            "variant": "h5",
          },
          "ref": null,
          "rendered": "All Products",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <div>
                <WithStyles(Typography)
                  id="product_total"
                  variant="caption"
                >
                  1000 Products
                </WithStyles(Typography)>
              </div>,
              <div>
                <WithStyles(Select)
                  disableUnderline={true}
                  name="limit"
                  onChange={[Function]}
                  value={8}
                >
                  <WithStyles(MenuItem)
                    value={8}
                  >
                    8 per page
                  </WithStyles(MenuItem)>
                  <WithStyles(MenuItem)
                    value={16}
                  >
                    16 per page
                  </WithStyles(MenuItem)>
                  <WithStyles(MenuItem)
                    value={20}
                  >
                    20 per page
                  </WithStyles(MenuItem)>
                </WithStyles(Select)>
              </div>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <WithStyles(Typography)
                  id="product_total"
                  variant="caption"
                >
                  1000 Products
                </WithStyles(Typography)>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "1000 Products",
                  "id": "product_total",
                  "variant": "caption",
                },
                "ref": null,
                "rendered": "1000 Products",
                "type": [Function],
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <WithStyles(Select)
                  disableUnderline={true}
                  name="limit"
                  onChange={[Function]}
                  value={8}
                >
                  <WithStyles(MenuItem)
                    value={8}
                  >
                    8 per page
                  </WithStyles(MenuItem)>
                  <WithStyles(MenuItem)
                    value={16}
                  >
                    16 per page
                  </WithStyles(MenuItem)>
                  <WithStyles(MenuItem)
                    value={20}
                  >
                    20 per page
                  </WithStyles(MenuItem)>
                </WithStyles(Select)>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": Array [
                    <WithStyles(MenuItem)
                      value={8}
                    >
                      8 per page
                    </WithStyles(MenuItem)>,
                    <WithStyles(MenuItem)
                      value={16}
                    >
                      16 per page
                    </WithStyles(MenuItem)>,
                    <WithStyles(MenuItem)
                      value={20}
                    >
                      20 per page
                    </WithStyles(MenuItem)>,
                  ],
                  "disableUnderline": true,
                  "name": "limit",
                  "onChange": [Function],
                  "value": 8,
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "children": "8 per page",
                      "value": 8,
                    },
                    "ref": null,
                    "rendered": "8 per page",
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "children": "16 per page",
                      "value": 16,
                    },
                    "ref": null,
                    "rendered": "16 per page",
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "children": "20 per page",
                      "value": 20,
                    },
                    "ref": null,
                    "rendered": "20 per page",
                    "type": [Function],
                  },
                ],
                "type": [Function],
              },
              "type": "div",
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": undefined,
            "componentStyle": ComponentStyle {
              "componentId": "sc-bwzfXH",
              "isStatic": true,
              "rules": Array [
                "
  display: flex;
  justify-content: space-between;
  align-items: center;
",
              ],
            },
            "displayName": "styled.div",
            "render": [Function],
            "styledComponentId": "sc-bwzfXH",
            "target": "div",
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "weight": 1,
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": undefined,
            "componentStyle": ComponentStyle {
              "componentId": "sc-bxivhb",
              "isStatic": false,
              "rules": Array [
                "
  border: ",
                [Function],
                "px solid #eeeeee;
",
              ],
            },
            "displayName": "styled.div",
            "render": [Function],
            "styledComponentId": "sc-bxivhb",
            "target": "div",
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [],
          },
          "ref": null,
          "rendered": Array [],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "height": 30,
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": undefined,
            "componentStyle": ComponentStyle {
              "componentId": "sc-htpNat",
              "isStatic": false,
              "rules": Array [
                "
  height: ",
                [Function],
                "px;
",
              ],
            },
            "displayName": "styled.div",
            "render": [Function],
            "styledComponentId": "sc-htpNat",
            "target": "div",
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <WithStyles(Pagination)
              classes={Object {}}
              currentPageColor="inherit"
              limit={8}
              nextPageLabel="Next page >"
              offset={0}
              onClick={[Function]}
              otherPageColor="inherit"
              previousPageLabel="< Previous page"
              reduced={true}
              total={1000}
            />,
            "end": "xs",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "classes": Object {},
              "currentPageColor": "inherit",
              "limit": 8,
              "nextPageLabel": "Next page >",
              "offset": 0,
              "onClick": [Function],
              "otherPageColor": "inherit",
              "previousPageLabel": "< Previous page",
              "reduced": true,
              "total": 1000,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
